### PR TITLE
Document AllowRecompilingViewsOnFileChange

### LINKED ISSUE #10108: Document AllowRecompilingViewsOnFileChange
[@Rick-Anderson, you worked on these aspects already (Razor SDK), so it's prob best if you take this, but re-assign me if you want me to do it. I've assembled all the relevant tidbits below. I'll leave the Sprint and Milestone in case you re-assign back to me&mdash;I was planning on working it very quickly when I set up this issue.]

Doc this where we discuss recompilation of views (e.g., https://docs.microsoft.com/aspnet/core/mvc/views/view-compilation).

> MVC still watches files for applications that don't have any views on disk. It should be possible to turn this behavior off completely when the application doesn't expect views on disk to override precompiled ones.
>
> There are some other benefits to adding this feature:
>
> * File watching for views is off (which seems to cause issues on certain OSes)
> * We do less work at runtime and there's no need to allocate change tokens for every view since nothing can change.

> ... in my current implementation (#8369), all I do is turn off file watching. Views still get compiled at runtime, they don't recompile when the file changes.

> We have disabled view recompilation when .cshtml change by default (except for the Development environment). A new flag called `AllowRecompilingViewsOnFileChange` has been introduced in `RazorViewEngineOptions`, which can be used to configure whether the view engine will watch for file changes. The settings is set to `false` by default for all environments but Development. If you are using Visual Studio to debug, then it will launch the application in the Development environment by default. You can find more details about the change by looking at this PR: https://github.com/aspnet/Mvc/pull/8369/files
>
> In case you would like to keep the view recompilation enabled for your project, you can choose so by configure Razor options in Startup.ConfigureService method using the following code block:

```csharp
services.AddMvc()
    .SetCompatibilityVersion(CompatibilityVersion.Version_2_2)
    .AddRazorOptions(options => options.AllowRecompilingViewsOnFileChange = true);
```

Code comments provide additional info :point_right: https://github.com/aspnet/AspNetCore/blob/release/2.2/src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngineOptions.cs#L198-L243

Will be removed for 3.0 :point_right: https://github.com/aspnet/AspNetCore/commit/f80490f99d4e38e1630e8cce739daf1b3a2b96a8 :point_right: The approach for 3.0 is already documented for Razor.SDK at https://docs.microsoft.com/aspnet/core/razor-pages/sdk

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 46a10a5c-bacc-1bcf-a7ae-499715e1bfdd
* Version Independent ID: 1585163b-b1d7-eeb8-ebc6-adfc2730f393
* Content: [Migrate from ASP.NET Core 2.1 to 2.2](https://docs.microsoft.com/en-us/aspnet/core/migration/21-to-22?view=aspnetcore-2.2&tabs=visual-studio)
* Content Source: [aspnetcore/migration/21-to-22.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/migration/21-to-22.md)
* Product: **aspnet-core**
* GitHub Login: @scottaddie
* Microsoft Alias: **scaddie**

### LINKED ISSUE #10180: Initial Migration: Build failed.
When I run "Update-Database' I get "Build failed."
Should I have not updated my csproj TargetFramework to 2.2
and the other versions to 2.2.0:

&lt;Project Sdk="Microsoft.NET.Sdk.Web"&gt;

  &lt;PropertyGroup&gt;
    &lt;TargetFramework&gt;netcoreapp2.2&lt;/TargetFramework&gt;
    &lt;AspNetCoreHostingModel&gt;InProcess&lt;/AspNetCoreHostingModel&gt;
  &lt;/PropertyGroup&gt;

  &lt;ItemGroup&gt;
    &lt;PackageReference Include="Microsoft.AspNetCore.App" /&gt;
    &lt;PackageReference Include="Microsoft.AspNetCore.Razor.Design" Version="2.2.0" PrivateAssets="All" /&gt;
    &lt;PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="2.2.0" /&gt;
  &lt;/ItemGroup&gt;

&lt;/Project&gt;

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 6719f08e-3bd7-dc1a-71df-f2ef9fbca9d8
* Version Independent ID: 7096fdb3-612e-9e00-bd0b-8ea4886a09ce
* Content: [Add a model to a Razor Pages app in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/model?view=aspnetcore-2.2&tabs=visual-studio)
* Content Source: [aspnetcore/tutorials/razor-pages/model.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/tutorials/razor-pages/model.md)
* Product: **aspnet-core**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #10108

* Taken just about verbatim from the [API reference](https://github.com/aspnet/AspNetCore/blob/release/2.2/src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngineOptions.cs#L198-L243) ... don't think we need engineering for this one.
* Note that I can't link 2.2 API until that issue is fixed, so I code fence the bits here. I'm noting on my 2.2 tracking issue to hit this up with proper API links later.

### CODE REVIEW

File: aspnetcore/mvc/views/view-compilation.md [hunk 1]
```diff
@@ -90,6 +90,25 @@ A *<project_name>.PrecompiledViews.dll* file, containing the compiled Razor file
 
 ::: moniker-end
 
+## Recompile views on file change behavior
+
+The <xref:Microsoft.AspNetCore.Mvc.Razor.RazorViewEngineOptions> `AllowRecompilingViewsOnFileChange` gets or sets a value that determines if Razor files (Razor Views and Razor Pages) are recompiled and updated if files change on disk.
+
+When set to `true`, MVC uses [IFileProvider.Watch](xref:<xref:Microsoft.Extensions.FileProviders.IFileProvider.Watch*>) to watch for changes to Razor files in configured <xref:Microsoft.Extensions.FileProviders.IFileProvider> instances.
```
Comment (Rick-Anderson) [inline aspnetcore/mvc/views/view-compilation.md]: @tdykstra 	@scottaddie 	should this be

When set to `true`, the ASP.NET Core runtime uses ...

Comment (tdykstra) [reply aspnetcore/mvc/views/view-compilation.md]: "MVC uses" is probably OK if it's really the MVC framework doing the watching -- if not, I would prefer your proposal.

Comment (scottaddie) [reply aspnetcore/mvc/views/view-compilation.md]: I suggest the following:

When set to `true`, ASP.NET Core uses...

Comment (guardrex) [reply aspnetcore/mvc/views/view-compilation.md]: I have another idear ... check the next commit.

**Comment by guardrex** [comment]
`Watch` watches :eyes: lol

File: aspnetcore/mvc/views/view-compilation.md [hunk 2]
```diff
@@ -90,6 +90,25 @@ A *<project_name>.PrecompiledViews.dll* file, containing the compiled Razor file
 
 ::: moniker-end
 
+## Recompile Razor files on change
+
+The <xref:Microsoft.AspNetCore.Mvc.Razor.RazorViewEngineOptions> `AllowRecompilingViewsOnFileChange` gets or sets a value that determines if Razor files (Razor Views and Razor Pages) are recompiled and updated if files change on disk.
```
Comment (scottaddie) [inline aspnetcore/mvc/views/view-compilation.md]:
```suggestion
The <xref:Microsoft.AspNetCore.Mvc.Razor.RazorViewEngineOptions> `AllowRecompilingViewsOnFileChange` gets or sets a value that determines if Razor files (Razor views and Razor Pages) are recompiled and updated if files change on disk.
```

**Comment by guardrex** [comment]
@scottaddie We shouldn't trust the engineers on capitalization? :smile: (I've gone right into *SillyRex* mode.)

**Comment by guardrex** [comment]
@Rick-Anderson U good? This one is ready to 🎷 *JAM* 🎸.

**Pull request merged**